### DISCUSSION TITLE
Maintain same BT ID with android root?

### DISCUSSION THREAD

**Discussion by chreds** [cagnulein/qdomyos-zwift#1758] [Q&A] [answered]
Can a rooted android device maintain the same Bluetooth id or is that not implemented?

Thanks.

**Comment by cagnulein** [comment]
I didn't implement it (nobody ask me before). do you have any documentation
about it? maybe it's something that you can do at system level

Il giorno mer 1 nov 2023 alle 15:01 chreds ***@***.***> ha
scritto:

**Comment by chreds** [comment, marked as answer]
Looks like one would have to create a custom ROM with:

#define BLE_LOCAL_PRIVACY_ENABLED FALSE

https://android.googlesource.com/platform//system/bt/+/95bb8d8eb533b5ddbb67bf6dd0b6e9d3c523f8ce/internal_include/bt_target.h

So probably no change required on your side.

**Reply by chreds** [reply]
On LineageOS looks like it's here if someone ever wanted to do this for their own device:
https://github.com/search?q=repo%3ALineageOS%2Fandroid_vendor_qcom_opensource_system_bt%20BLE_LOCAL_PRIVACY_ENABLED&type=code

**Reply by chreds** [reply]
Note they switched bluetooth drivers in Android 13 and I can't find the flag to disable BLE privacy. I'm going to compile LineageOS 19.1 now with this flag disabled and see if that actually works how I want it to.

**Comment by cagnulein** [comment]
Very cool!

Il giorno mer 1 nov 2023 alle 16:52 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
Yeah unfortunately it's not a runtime config option.

**Comment by cagnulein** [comment]
Ok, let me know.

Il giorno mar 21 nov 2023 alle 19:58 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
@cagnulein I recompiled Android (Lineage 18.1) with the BLE_LOCAL_PRIVACY_ENABLED flag turned off. Now I don't see QZ showing up in Rouvy at all. Does this debug shed any light?
[debug-Mon_Dec_4_18_36_58_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13555702/debug-Mon_Dec_4_18_36_58_2023.log)

**Comment by cagnulein** [comment]
from the qz side everything is ok, for rouvy I guess you have to enable the
bike power sensor in the experimental settings.
Let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 5 dic 2023 alle ore 06:45 chreds ***@***.***>
ha scritto:

**Comment by chreds** [comment]
Ok I got the bt id staying the same with custom lineageos. I had to use lineage 18 since newer android has a different Bluetooth stack where that privacy flag doesn't exist anymore. Everything is working perfect now.

How do I disable qz from sending heart rate info? I use my watch for that connected via ant+ to the bike.

**Reply by cagnulein** [reply]
that's great! can you share this build somehow? there is no way to disable qz as an heart rate service from the settings, but why you need this? it shouldn't break anything.
Let me know

**Comment by chreds** [comment]
It would just simplify connection since Rouvy auto-connects to it. Not a
big deal I can switch to the watch pretty easy.

The build I made is only built for Pixel 4a so not sure how useful it would
be for people. If anyone else has one of those I'm happy to share the build
with them.

I followed these instructions:
Build for sunfish | LineageOS Wiki
<https://wiki.lineageos.org/devices/sunfish/build>

Hardest part is to find existing lineage 18 builds for your devices to
build the proprietary blobs during the lineage source build since they only
publish the latest 7 builds or something.

I flashed Google Factory Android 11 build to get Android 11 firmwares then
flashed  [ROM][UNOFFICIAL][SUNFISH][A11] LineageOS 18.1 | XDA Forums
<https://xdaforums.com/t/rom-unofficial-sunfish-a11-lineageos-18-1.4481023/> to
build proprietary blobs then flashed my modified 18.1 build with the
BLE Privacy turned off.

I modified bt_include.h #define BLE_LOCAL_PRIVACY_ENABLED FALSE before the
build.

**Reply by cagnulein** [reply]
this is not useful only for rouvy but even for garmin devices. i guess you should now able to send metrics to a garmin device by bluetooth. do you have one?

**Reply by chreds** [reply]
I do have one. What did you want me to try?

**Comment by cagnulein** [comment]
check for example if you can pair your phone via bluetooth to the garmin.
you may need to enable the virtual device heart only setting in the
experimental settings just to force qz to send only hr to the garmin, so it
will be easy to check this
thanks

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 15 dic 2023 alle 17:57 chreds ***@***.***> ha
scritto:
